### PR TITLE
Improve network interface support for LHOST (OptAddressLocal)

### PR DESCRIPTION
**Pull request by wvu**
# ~WIP~ Test me!

~Do we really want to exclude loopback? It's occasionally useful, and there's already a warning when setting it.~ No longer excluded.

- [x] Test tab completion of IPv4 and IPv6 by address
- [x] Note that the full list of addresses is displayed now (fixed regression)
- [x] Test tab completion of IPv4 and IPv6 by interface
- [x] Test setting of IPv4 and IPv6 by address
- [x] Test setting of IPv4 and IPv6 by interface

#8336, #10086

### CODE REVIEW

File: lib/msf/core/opt_address_local.rb [hunk 1]
```diff
@@ -9,32 +9,36 @@ module Msf
 #
 ###
 class OptAddressLocal < OptAddress
+  def interfaces
+    NetworkInterface.interfaces || []
+  end
+
   def normalize(value)
```
Comment (bcoles) [inline lib/msf/core/opt_address_local.rb]:
```ruby
  def normalize(value)
    return unless value.kind_of?(String)
    return value unless interfaces.include?(value)

    ip_address = NetworkInterface.addresses(value).values.flatten.map{|x| x['addr']}.select do |addr|
      begin
        IPAddr.new(addr).ipv4?
      rescue IPAddr::InvalidAddressError
        nil
      end
    end

    return if ip_address.blank?

    ip_address.first
  end
```

**Comment by jmartin-r7** [comment]
So pretty :+1:
```
msf5 > use exploit/multi/handler
msf5 exploit(multi/handler) > set payload linux/x64/meterpreter_reverse_tcp
payload => linux/x64/meterpreter_reverse_tcp
msf5 exploit(multi/handler) > set LHOST
set LHOST 127.0.0.1                        set LHOST bridge0                          set LHOST gif0
set LHOST X.X.X.X                          set LHOST en0                              set LHOST lo0
set LHOST X.X.X.X                          set LHOST en1                              set LHOST p2p0
set LHOST X.X.X.X                          set LHOST en2                              set LHOST stf0
set LHOST X.X.X.X                          set LHOST xxxx::x%lo0                      set LHOST utun0
set LHOST ::1                              set LHOST xxxx::xxxx:xxxx:xxxx:xxxx%en0    set LHOST utun1
set LHOST XHC20                            set LHOST xxxx::xxxx:xxxx:xxxx:xxxx%utun0  set LHOST vmnet1
set LHOST awdl0                            set LHOST xxxx::xxxx:xxxx:xxxx:xxxx%awdl0  set LHOST vmnet8
msf5 exploit(multi/handler) > set LHOST
```

**Comment by bcoles** [comment]
```
msf5 exploit(multi/handler) > set lhost 
set lhost 127.0.0.1                      set lhost eth0
set lhost 172.16.X.X                     set lhost XXXX::XXX:XXXX:XXXX:XXXX%eth0
set lhost ::1                            set lhost lo
```

**Comment by h00die** [comment]
This looks the hotness

**Comment by busterb** [comment]
LGTM too, thanks @wvu-r7

File: lib/msf/core/opt.rb [hunk 2]
```diff
@@ -40,7 +40,7 @@ def self.Proxies(default=nil, required=false, desc="A proxy chain of format type
       Msf::OptString.new(__method__.to_s, [ required, desc, default ])
     end
 
-    # @return [OptAddress]
+    # @return [OptAddressRange]
```
Comment (busterb) [inline lib/msf/core/opt.rb]: conflict, I fixed this too :)

**Comment by busterb** [comment]
## Release Notes

This implements support for tab-completing network interface names for the LHOST parameter.

**Pull request merged**